src/history/hs.c

```diff
@@ -425,38 +426,17 @@ __hs_insert_record_with_btree_int(WT_SESSION_IMPL *session, WT_CURSOR *cursor, W
     WT_STAT_CONN_INCR(session, cache_hs_insert);
 
 err:
-    if (ret != 0)
+    if (ret != 0) {
         __wt_free_update_list(session, &hs_upd);
-    /*
-     * If we inserted an update with no timestamp, we need to delete all history records for that
-     * key that are further in the history table than us (the key is lexicographically greater). For
-     * timestamped tables that are occasionally getting a non-timestamped update, that means that
-     * all timestamped updates should get removed. In the case of non-timestamped tables, that means
-     * that all updates with higher transaction ids will get removed (which could happen at some
-     * more relaxed isolation levels).
-     */
-    if (ret == 0 && upd->start_ts == WT_TS_NONE) {
-#ifdef HAVE_DIAGNOSTIC
+
         /*
-         * We need to initialize the last searched key so that we can do key comparisons when we
-         * begin iterating over the history store. This needs to be done otherwise the subsequent
-         * "next" calls will blow up.
+         * We did a row search, release the cursor so that the page doesn't continue being held.
+         *
+         * If we were successful, do NOT reset the cursor. We may want to make use of its position
+         * later to remove timestamped entries.
          */
-        WT_TRET(__wt_cursor_key_order_init(cbt));
-#endif
-        F_SET(cursor, WT_CURSTD_IGNORE_TOMBSTONE);
-        /* We're pointing at the newly inserted update. Iterate once more to avoid deleting it. */
-        ret = cursor->next(cursor);
-        if (ret == WT_NOTFOUND)
-            ret = 0;
-        else if (ret == 0) {
-            WT_TRET(__hs_delete_key_from_pos(session, cursor, btree->id, key));
```

Comment:
@agorrod 
This is the path I'm referring to. If we insert successfully above and then get `WT_RESTART` here, we'll repeat the entire function body again including the initial insert.